README.md
When this module is active, all full-text searches of your [ApostropheCMS](https://apostrophecms.org) site run through [Elasticsearch](https://www.elastic.co/products/elasticsearch), replacing the built-in MongoDB text search engine normally used by Apostrophe.

**You must have an Elasticsearch server.** By default, Elasticsearch is looked for on the same computer, but you may specify `elasticsearchOptions` to [change that](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/configuration.html#config-options).

```
npm install apostrophe-elasticsearch
```

```javascript
// in app.js
modules: {
  'apostrophe-elasticsearch': {}
}
```

You can also configure custom boosts for various properties if present in the doc, decide what fields to index, and do much more, including passing options through to Elasticsearch. Here is a more complex configuration with all of the options in play:

```javascript
// in app.js
modules: {
  'apostrophe-elasticsearch': {
    // Default matches your shortName. Index names
    // will look like: myprojectaposdocsen (for en locale),
    // myprojectaposdocsfr (for fr locale), etc.
    baseName: 'myproject',

    // elasticsearch config options, please see:
    // https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/configuration.html#config-options
    elasticsearchOptions: {
      // This is the default
      host: 'localhost:9200'
    },

    // Add an extra field to be indexed directly by Elasticsearch.
    // Certain fields are always indexed in addition to these,
    // including lowSearchText and highSearchText which most likely
    // include your schema field's text already

    addFields: [ 'interests' ],

    // Relative importance. Note that if you don't specify elasticsearch
    // does a rather good job figuring this out on its own.

    boosts: {
      tags: 50,
      title: 100,
      highSearchText: 10,
      lowSearchText: 1
    },

    // Simple shortcut to set the analyzer for all indexes.
    // Useful for single-language sites
    analyzer: 'french',

    // Shortcut to specify analyzers by locale (with apostrophe-workflow)
    analyzers: {
      master: 'french',
      fr: 'french',
      en: 'english'
    },

    // Becomes the elasticsearch `settings` property of each index
    indexSettings: {
      'analysis': {
        'analyzer': 'french'
      }
    },
    
    // Sets additional subproperties of the elasticsearch `settings` property
    // of the index, on a per-locale basis (for use with apostrophe-workflow)
    localeIndexSettings: {
      'en': {
        'analysis': {
          'analyzer': 'english'
        }
      }
    },
  }
}
```

> **"What fields are always indexed?"** `lowSearchText` contains all of the text of the doc, including rich text editor content, stripped of its markup. `highSearchText` contains only text in `string` and `tags` schema fields and is often boosted higher. Note that these will both contain `title`. Also, `title`, `slug`, `path`, `tags` and `type` are always indexed. If you really want to, you can use the `fields` option to override this base list of fields and specify exactly the array of fields you want. However, certain fields are mandatory for the query engine to operate correctly and will automatically be added back.

> **"What are the possible properties for `indexSettings` and `localeIndexSettings`?"** See the [elasticsearch documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules.html).

Now we need to index our existing docs in Elasticsearch. You must run this task at least once when adding this module:

```
node app apostrophe-elasticsearch:reindex --verbose
```

This task can take a long time on a large existing database, so the `--verbose` option displays progress information periodically, including a (somewhat optimistic) estimate of the time remaining. If you prefer you can leave it off for silent operation.

> **Apostrophe automatically updates the index as you edit docs.** You don't have to run this task all the time! However, if you change your `fields` option, change `indexSettings`, `analyzer`, `analyzers` or `localeIndexSettings`, or make significant edits that are invisible to Apostrophe via direct MongoDB updates, you should run this task again. You do not have to run it again when you change `boosts`.

Apostrophe will now use Elasticsearch to implement all searches that formerly used the built-in MongoDB text index:

* Sitewide search
* Anything else based on the `search()` cursor filter
* Anything else based on the `autocomplete()`  cursor filter
* Anything based on the `q` or `search` query parameters to an apostrophe-pieces-page (these invoke `search` via `queryToFilters`)
* Anything based on the `autocomplete` query parameter (this invokes `autocomplete` via `queryToFilters`)

> **All queries involving `search` or `autocomplete` will always sort their results in the order returned by Elasticsearch,** regardless of the `sort` cursor filter. Relevance is almost always the best sort order for results that include a free text search, and to do otherwise would require an exhaustive search of every match in Elasticsearch (potentially thousands of docs), just in case the last one scores higher according to some other criterion. Support for sorting on some other properties may be added later as more information is mirrored in Elasticsearch.

> **If a query cannot be completely satisfied by the first 10,000 matches in Elasticsearch, the results may be incomplete.** This is true even if the first 10,000 are simply being bypassed with `skip`. This is a built-in safeguard in Elasticsearch due to the memory and time impact of such searches. In rare cases, such as a search for a word mentioned in one document that is less relevant than the first 10,000 matching documents but is still the only relevant document according to MongoDB criteria also present in your query, this could impact the completeness of the response. However certain common categories of MongoDB query criteria are in fact replicated to Elasticsearch to mitigate this problem and improve performance (see below).

## Theory of operation

MongoDB and Elasticsearch both support many types of queries, including ranges, exact matches, and fuzzy matches. However, they are not identical query languages. Not every property is necessarily appropriate to copy to Elasticsearch. And Apostrophe queries that include text searches can also include arbitrarily complex MongoDB criteria on other properties.

For these reasons, we should not attempt to translate 100% of queries directly to Elasticsearch. However, we must not provide the user with results they should not see.

To meet both criteria, queries flow like this. For purposes of this discussion, `search` and `autocomplete` are equivalent.

* `search` is not present: query flows through the normal MongoDB query path and the remainder of these steps are not followed.
* `search` is present: the text search query parameter is given to Elasticsearch. In addition, a subset of additional criteria present in the query are given to Elasticsearch to narrow the results. Specifically, code inspired by that in [apostrophe-optimizer](https://npmjs.org/package/apostrophe-optimizer) is used to locate certain common "hard constraints" on the query, such as `type`, `tags` and `_id`. These are used to build a `filter` clause which does not affect relevance ranking. This improves performance by preventing Elasticsearch from exhaustively searching blog posts when a query is restricted to events in any case. Elasticsearch is asked for an initial batch of results, rather than all possible results.
* The results of this stage are fed through the normal MongoDB query path, without `$text` clauses, to ensure that additional MongoDB criteria present, such as permissions, are fully implemented.
* Additional batches of results are retrieved from the Elasticsearch stage and fed through the MongoDB query stage until the `limit` of the query is satisfied (if any) or no more results are obtained.
* Actual results are supplied to the caller via a wrapper that reimplements Apostrophe's `lowLevelMongoCursor` method for this use case.
* In this scheme, implementation of `toCount` requires an exhaustive query, however the projection is restricted to the `_id` only in both Elasticsearch and MongoDB to mitigate the impact.
* In this scheme, implementation of `toDistinct` also requires an exhaustive query, however the projection is limited to the property in question to mitigate the impact.
* `apostrophe-workflow` is present: for performance, and to avoid complex filter queries, a separate elasticsearch index is used for each locale. Documents that are not locale-specific are replicated to the indexes for all of the locales. 

This algorithm implements a "cross-domain join" between two different data stores, MongoDB and Elasticsearch. This is an inherently difficult problem to optimize completely, however the communication of a subset of the MongoDB criteria to Elasticsearch eliminates much of the overhead without the need to implement 100% of the MongoDB query language on top of Elasticsearch with perfect fidelity.


CHANGELOG.md
# Changelog

## 2.1.3

Better test for inappropriate values given to the `search` cursor filter method, with full unit testing.

## 2.1.2

Does not crash if `null` or `undefined` is passed to the `search` cursor filter method. Thanks to giuseppecm.

## 2.1.1

Fixed an oversight in the documentation which did not explain that the use of `addFields` is preferred over specifying `fields` in its entirety.

Also made sure the set of fields that are directly queried for exact matches to speed up queries are always indexed, even if they are left out of the `fields` option.

If you received no search results due to a combination of these issues, run the reindex task again and try those searches again.

## 2.1.0

Major performance and reliability improvements to the indexing procedure, especially the `apostrophe-elasticsearch:reindex` task. Batch operations are now used for better throughput, and when reindexing one locale is processed at a time to minimize resource contention between the different indexes. In addition, the reindex task now has a `--verbose` option that displays progress information, including an estimate of the remaining time. This estimate tends to be a bit optimistic because Elasticsearch does slow down somewhat as it fills up.

## 2.0.1

Documentation tweaks.

## 2.0.0

Initial release, with passing tests.


test/test.js
const assert = require('assert');
const _ = require('lodash');

describe('apostrophe-elasticsearch: ', function() {

  let apos = null;

  this.timeout(20000);

  after(function(done) {
    try {
      require('apostrophe/test-lib/util').destroy(apos, done);
    } catch (e) {
      console.warn('Old version of apostrophe does not export test-lib/util library, just dropping old test db');
      apos.db.dropDatabase();
      setTimeout(done, 1000);
    }
  });

  it('should be a property of the apos object', function(done) {
    apos = require('apostrophe')({
      testModule: true,
      baseUrl: 'http://localhost:7780',
      modules: {
        'apostrophe-express': {
          port: 7780
        },
        'apostrophe-elasticsearch': {
        },
        'apostrophe-pages': {
          park: [
            {
              title: 'Tab One',
              type: 'default',
              slug: '/tab-one',
              _children: [
                {
                  title: 'Tab One Child One',
                  type: 'default',
                  slug: '/tab-one/child-one'
                },
                {
                  title: 'Tab One Child Two',
                  type: 'default',
                  slug: '/tab-one/child-two'
                }
              ]
            },
            {
              title: 'Tab Two',
              type: 'default',
              slug: '/tab-two',
              _children: [
                {
                  title: 'Tab Two Child One',
                  type: 'default',
                  slug: '/tab-two/child-one'
                },
                {
                  title: 'Tab Two Child Two',
                  type: 'default',
                  slug: '/tab-two/child-two'
                }
              ]
            },
            {
              title: 'Products',
              type: 'products-page',
              slug: '/products'
            }
          ],
          types: [
            {
              name: 'home',
              label: 'Home'
            },
            {
              name: 'default',
              label: 'Default'
            },
            {
              name: 'products',
              label: 'Products'
            }
          ]
        },
        'products': {
          extend: 'apostrophe-pieces',
          name: 'product',
          alias: 'products',
          construct: function(self, options) {
            self.generate = function(i) {
              let piece = self.newInstance();
              // one word for ease of search test
              piece.title = 'product' + (i + 1);
              piece.age = i % 10;
              piece.body = {
                type: 'area',
                items: [
                  {
                    _id: apos.utils.generateId(),
                    type: 'apostrophe-rich-text',
                    // everybody gets calligraphy, 50% get stenography
                    content: '<h4>calligraphy ' + ((i & 1) ? 'stenography' : '') + '</h4>'
                  }
                ]
              };
              piece.published = true;
              return piece;
            };
          }
        },
        'products-pages': {
          extend: 'apostrophe-pieces-pages'
        }
      },
      afterInit: function(callback) {
        assert(apos.modules['apostrophe-elasticsearch']);
        assert(apos.modules['apostrophe-elasticsearch'].refresh);
        return callback(null);
      },
      afterListen: function(err) {
        assert(!err);
        done();
      }
    });
  });

  it('reindex to remove old test cases', function() {
    return apos.tasks.invoke('apostrophe-elasticsearch:reindex', { verbose: true });
  });

  it('insert many products for test purposes', function() {
    return apos.tasks.invoke('products:generate', { total: 100 });
  });

  it('insert an unpublished product for test purposes', function() {
    let product = _.assign(apos.modules.products.newInstance(), {
      title: 'Unpublished',
      slug: 'unpublished',
      published: false
    });
    return apos.modules.products.insert(apos.tasks.getReq(), product);
  });

  it('search for one product', function() {
    const req = apos.tasks.getAnonReq();
    const cursor = apos.products.find(req, {}).search('product50');
    return cursor.toArray().then(function(products) {
      assert(products);
      assert(products.length === 1);
      assert(products[0].title === 'product50');
      // Make sure elasticsearch is in use
      assert(cursor.get('elasticsearch'));
    });
  });

  it('no search (return everything) if search parameter is null', function() {
    const req = apos.tasks.getAnonReq();
    const cursor = apos.products.find(req, {}).search(null);
    return cursor.toArray().then(function(products) {
      assert(products);
      assert(products.length === 100);
      // Make sure elasticsearch is not in use
      assert(!cursor.get('elasticsearch'));
    });
  });

  it('no search (return everything) if search parameter is undefined', function() {
    const req = apos.tasks.getAnonReq();
    const cursor = apos.products.find(req, {}).search(undefined);
    return cursor.toArray().then(function(products) {
      assert(products);
      assert(products.length === 100);
      // Make sure elasticsearch is not in use
      assert(!cursor.get('elasticsearch'));
    });
  });

  it('no search (return everything) if search parameter is empty string', function() {
    const req = apos.tasks.getAnonReq();
    const cursor = apos.products.find(req, {}).search('');
    return cursor.toArray().then(function(products) {
      assert(products);
      assert(products.length === 100);
      // Make sure elasticsearch is not in use
      assert(!cursor.get('elasticsearch'));
    });
  });

  it('search for unpublished product should fail', function() {
    const req = apos.tasks.getAnonReq();
    return apos.products.find(req, {}).search('unpublished').toArray().then(function(products) {
      assert(products);
      assert(products.length === 0);
    });
  });

  it('search for unpublished product should fail with published(false) if we lack the permissions', function() {
    const req = apos.tasks.getAnonReq();
    return apos.products.find(req, {}).search('unpublished').published(false).toArray().then(function(products) {
      assert(products);
      assert(products.length === 0);
    });
  });

  it('search for unpublished product should succeed with published(false) if we have the permissions', function() {
    const req = apos.tasks.getReq();
    return apos.products.find(req, {}).search('unpublished').published(false).toArray().then(function(products) {
      assert(products);
      assert(products.length === 1);
      assert(products[0].published === false);
      assert(products[0].title === 'Unpublished');
    });
  });

  it('can get everything', function() {
    const req = apos.tasks.getAnonReq();
    return apos.products.find(req).search('calligraphy').toArray().then(function(docs) {
      assert(docs);
      assert(docs.length === 100);
    });
  });

  it('can get 50%', function() {
    const req = apos.tasks.getAnonReq();
    return apos.products.find(req).search('stenography').toArray().then(function(docs) {
      assert(docs);
      assert(docs.length === 50);
    });
  });

  it('can paginate the 50% properly', function() {
    const req = apos.tasks.getAnonReq();
    let i = 0;
    let titlesSeen = {};
    return testNextBatch();
    function testNextBatch() {
      return apos.products.find(req).search('stenography').skip(i).limit(5).toArray().then(function(docs) {
        assert(docs);
        assert(docs.length === 5);
        _.each(docs, function(doc) {
          assert(!titlesSeen[doc.title]);
          titlesSeen[doc.title] = true;
        });
        i += 5;
        if (i === 50) {
          assert(_.keys(titlesSeen).length === 50);
          return;
        }
        return testNextBatch();
      });
    }
  });

  it('unrelated queries work properly', function() {
    const req = apos.tasks.getAnonReq();
    return apos.products.find(req, { title: 'product50' }).toObject().then(function(product) {
      assert(product);
      assert(product.title === 'product50');
    });
  });

  it('distinct works properly', function() {
    const req = apos.tasks.getReq();
    return apos.products.find(req).search('calligraphy').toDistinct('age').then(function(ages) {
      assert(ages);
      assert(ages.length === 10);
      for (let i = 0; (i < 10); i++) {
        assert(_.find(ages, function(age) {
          return age === i;
        }) !== undefined);
      }
    });
  });

  it('distinct works properly with counts', function() {
    const req = apos.tasks.getReq();
    const cursor = apos.products.find(req).search('calligraphy').distinctCounts(true);
    return cursor.toDistinct('age').then(function(ages) {
      assert(ages);
      assert(ages.length === 10);
      for (let i = 0; (i < 10); i++) {
        assert(_.find(ages, function(age) {
          return age === i;
        }) !== undefined);
      }
      let counts = cursor.get('distinctCounts');
      for (let i = 0; (i < 10); i++) {
        assert(counts[i] === 10);
      }
    });
  });

  it('safeQuery detects all arguments that do not convert to nonempty strings', function() {
    const module = apos.modules['apostrophe-elasticsearch'];
    const values = [ null, undefined, false, true, 0, '0', 1, '1', '' ];

    for (const s of values) {
      try {
        const string = s.toString();
        if (string.length) {
          assert(module.safeQuery(s));
        } else {
          assert(!module.safeQuery(s));
        }
      } catch (e) {
        assert(!module.safeQuery(s));
      }
    } 

  });

});


index.js
const elasticsearch = require('elasticsearch');
const async = require('async');
const _ = require('lodash');

module.exports = {
  afterConstruct: function(self, callback) {
    self.addReindexTask();
    return self.connect(callback);
  },
  construct: function(self, options) {
    self.fields = (self.options.fields || [ 'title', 'slug', 'path', 'tags', 'type', 'lowSearchText', 'highSearchText' ]).concat(self.options.addFields || []);
    // Mandatory because the query engine does exact matches on them
    // for performance
    const mandatoryFields = [ 'slug', 'path', 'type', 'tags' ];
    self.fields = self.fields.concat(mandatoryFields);
    self.fields = _.uniq(self.fields);
    self.apos.define('apostrophe-cursor', require('./lib/cursor.js'));
    self.connect = function(callback) {
      self.baseName = self.options.baseName || self.apos.shortName;
      // Index names are restricted to lowercase letters
      self.docIndex = self.baseName.toLowerCase() + 'aposdocs';
      let host = null;
      self.esOptions = (self.options.elasticsearchOptions || {});
      if (self.options.port) {
        host = (self.options.host || 'localhost') + ':' + self.options.port;
      } else {
        // Common convention with elasticsearch is one string with both host and port
        host = self.options.host || 'localhost:9200';
      }
      self.esOptions.host = self.esOptions.host || host;
      self.esOptions.apiVersion = self.esOptions.apiVersion || '6.4';

      self.client = new elasticsearch.Client(self.esOptions);
      return self.client.ping({
        requestTimeout: 5000
      }, callback);
    };

    self.docAfterSave = function(req, doc, options, callback) {
      const args = {
        body: self.getBulkIndexCommandsForDoc(req, doc, options),
        refresh: (!options.elasticsearchDefer).toString()
      };
      return self.client.bulk(args, callback);
    };

    // Return an array of parameters to `self.client.bulk` to
    // index the doc in question. There could be just one index
    // action description and one document body, or there could be
    // many such pairs, depending on workflow requirements.

    self.getBulkIndexCommandsForDoc = function(req, doc, options) {
      const commands = [];
      const body = {};
      _.each(self.fields, function(field) {
        const value = doc[field];
        let good = false;
        if (Array.isArray(value)) {
          if ((typeof value[0]) !== 'object') {
            good = true;
          }
        } else if ((typeof value) !== 'object') {
          good = true;
        }
        if (good) {
          body[field] = doc[field];
          // Allow exact matches of fields too without as much overthinking
          // of types, but don't redundantly store the 'exact' value where
          // it's too large to be requested that way anyway
          if (((!doc[field]) || JSON.stringify(doc[field]).length < 4096)) {
            body[field + 'ESExact'] = doc[field];
          }
        }
      });
      // No idea why it isn't cool to have this in the body too,
      // but ES flips out
      delete body._id;
      const locale = options.effectiveLocale || doc.workflowLocale;
      if (locale) {
        commands.push({
          index: {
            _index: self.getLocaleIndex(locale),
            _type: 'aposDoc',
            _id: doc._id
          }
        });
        commands.push(body);
      } else {
        // Not locale specific, so must appear in all locale indexes
        let locales = [ 'default' ];
        const workflow = self.apos.modules['apostrophe-workflow'];
        if (workflow) {
          locales = _.keys(workflow.locales);
        }
        _.each(locales, function(locale) {
          commands.push({
            index: {
              _index: self.getLocaleIndex(locale),
              _type: 'aposDoc',
              _id: doc._id
            }
          });
          commands.push(body);
        });
      }
      return commands;
    };

    self.indexNamesSeen = {};
    self.getLocaleIndex = function(locale) {
      locale = locale.toLowerCase();
      const indexName = self.docIndex + locale.replace(/[^a-z]/g, '');
      if (self.indexNamesSeen[indexName] && (self.indexNamesSeen[indexName] !== locale)) {
        throw new Error('apostrophe-elasticsearch: the locale names ' + locale + ' and ' + self.indexNamesSeen[indexName] + ' cannot be distinguished purely by their lowercased letters. elasticsearch allows no other characters in index names.');
      }
      self.indexNamesSeen[indexName] = locale;
      return indexName;
    };
    self.addReindexTask = function() {
      return self.addTask('reindex', 'Usage: node app apostrophe-elasticsearch:reindex\n\nYou should only need this task once under normal conditions.', self.reindexTask);
    };

    self.reindexTask = function(apos, argv, callback) {
      self.verbose('reindex task launched');
      return self.apos.locks.withLock('apostrophe-elasticsearch-reindex', function(callback) {
        return async.series([
          self.dropIndexes,
          self.createIndexes,
          self.index,
          self.refresh
        ], callback);
      });
    };

    // Drop all indexes. Called by the reindex task, otherwise not needed

    self.dropIndexes = function(callback) {
      self.verbose('dropping indexes');
      return self.client.cat.indices({
        h: ['index']
      }, function(err, result) {
        if (err) {
          return callback(err);
        }
        // Strange return format
        let indexes = (result || '').split(/\n/);
        indexes = _.filter(indexes, function(index) {
          return index.substr(0, self.docIndex.length) === self.docIndex;
        });
        if (!indexes.length) {
          return callback(null);
        }
        return self.client.indices.delete({
          index: indexes
        }, callback);
      });
    };

    self.getLocales = function() {
      const workflow = self.apos.modules['apostrophe-workflow'];
      return workflow ? _.keys(workflow.locales) : [ 'default' ];
    };

    // Create all indexes. Called by the reindex task, otherwise not needed
    self.createIndexes = function(callback) {
      self.verbose('creating indexes');
      const locales = self.getLocales();
      return async.eachSeries(locales, function(locale, callback) {
        const properties = {};
        _.each(self.fields, function(field) {
          properties[field] = {
            type: 'text'
          };
          properties[field + 'ESExact'] = {
            type: 'keyword'
          };
        });
        return self.client.indices.create({
          index: self.getLocaleIndex(locale),
          body: {
            settings: self.getLocaleSettings(locale),
            mappings: {
              aposDoc: {
                properties: properties
              }
            }
          }
        }, callback);
      }, callback);
    };

    self.getLocaleSettings = function(locale) {
      locale = locale.replace(/-draft$/, '');
      const settings = {};
      _.merge(settings, self.options.indexSettings || {});
      if (self.options.analyzer) {
        _.merge(settings, {
          'analysis': {
            'analyzer': self.options.analyzer
          }
        });
      }
      _.merge(settings, (self.options.localeIndexSettings || {})[locale] || {});
      if (self.options.analyzers && self.options.analyzers[locale]) {
        _.merge(settings, {
          'analysis': {
            'analyzer': self.options.analyzers[locale]
          }
        });
      }
      return settings;
    };

    // Index all documents. Called by the reindex task, otherwise not needed
    self.index = function(callback) {
      const batchSize = self.options.batchSize || 1000;
      const req = self.apos.tasks.getReq();
      self.verbose('Indexing all documents');
      const workflow = self.apos.modules['apostrophe-workflow'];
      const locales = (workflow && _.keys(workflow.locales)) || [ null ];
      const start = Date.now();
      let li = 0;
      return async.eachSeries(locales, processLocale, callback);

      function processLocale(locale, callback) {
        let reindexed = 0;
        let last = '';
        let count;

        self.verbose(locale + ': counting docs for progress display');
        return self.apos.docs.db.count({
          $or: [
            {
              workflowLocale: locale
            },
            {
              // docs with no locale must be indexed with every locale
              workflowLocale: null
            }
          ]
        }, function(err, _count) {
          if (err) {
            return callback(err);
          }
          count = _count;
          // Do the next batch, recursively
          return nextBatch(callback);
        });
        function nextBatch(callback) {
          // const start = Date.now();
          return self.apos.docs.db.find({
            $or: [
              {
                workflowLocale: locale
              },
              {
                // docs with no locale must be indexed with every locale
                workflowLocale: null
              }
            ],
            _id: { $gt: last }
          }).sort({ _id: 1 }).limit(batchSize).toArray(function(err, docs) {
            // const end = Date.now();
            if (err) {
              return callback(err);
            }
            if (!docs.length) {
              // This is how we terminate successfully
              li++;
              return callback(null);
            }
            const lastDoc = docs[docs.length - 1];
            last = lastDoc._id;
            // const iStart = Date.now();
            return indexBatch(docs, function(err) {
              // const iEnd = Date.now();
              if (err) {
                return callback(err);
              }
              // self.verbose('locale: ' + locale + ' last: ' + last + ' fetch: ' + (end - start) + ' index: ' + (iEnd - iStart));
              return nextBatch(callback);
            });
          });
        }
        function indexBatch(docs, callback) {
          reindexed += docs.length;
          const portion = (li / locales.length) + (reindexed / count * (1 / locales.length));
          const percent = Math.floor(portion * 100 * 100) / 100;
          const time = ((Date.now() - start) / portion);
          const remaining = time - (Date.now() - start);
          const localeLabel = locale ? `${locale}: ` : '';
          self.verbose(`${localeLabel}indexed ${reindexed} of ${count}, locale ${li + 1} of ${locales.length} (${percent}%) (${formatTime(remaining)} remaining)`);
          const args = {
            body: _.flatten(
              _.map(docs, function(doc) {
                return self.getBulkIndexCommandsForDoc(req, doc, {
                  effectiveLocale: locale
                });
              })
            )
          };
          return self.client.bulk(args, function(err) {
            if (err && (err.code === 'ECONNRESET')) {
              self.verbose('ECONNRESET, trying a new connection');
              self.client = new elasticsearch.Client(self.esOptions);
              return indexBatch(docs, callback);
            }
            if (err && self.errorIsSplittable(err) && (docs.length > 1)) {
              // Slice and dice until we get through
              self.verbose(err.statusCode + ' suggests too big, splitting up this batch, if you see this a lot reduce batchSize option');
              const pivot = Math.floor(docs.length / 2);
              const batches = [
                docs.slice(0, pivot),
                docs.slice(pivot)
              ];
              return async.eachSeries(batches, indexBatch, callback);
            }
            return callback(err);
          });
        }
        function formatTime(ms) {
          ms /= 1000;
          ms /= 60;
          ms = Math.floor(ms);
          const hours = Math.floor(ms / 60);
          const minutes = ms - (hours * 60);
          return `${hours}h${minutes}m`;
        }
      }

    };

    self.errorIsSplittable = function(err) {
      return (err.statusCode === 413) || (err.statusCode === 408);
    };

    // Refresh the index (commit it so it can be used). Called by the reindex task
    // once at the end for efficiency
    self.refresh = function(callback) {
      self.verbose('refreshing');
      const locales = self.getLocales();
      const indexes = _.map(locales, self.getLocaleIndex);
      return self.client.indices.refresh({ index: indexes }, callback);
    };

    self.verbose = function(s) {
      if (self.apos.argv.verbose || self.options.verbose) {
        self.apos.utils.info(self.__meta.name + ': ' + s);
      }
    };

    // Returns true if argument given converts to a nonempty string and
    // is not otherwise harmful to pass to elasticsearch (so far, just
    // the nonempty string check appears to be necessary)

    self.safeQuery = function(s) {
      return (s !== null) && (s !== undefined) && (s.toString().length);
    };
  }
};
